### PR TITLE
feat(columnar): delete-density rewrite, read-path perf, backpressure + zstd fixes

### LINKED ISSUE #551: feat(columnar): delete-bitmap density rewrite policy + public benchmark matrix
Two related columnar follow-ups, kept in one issue: a compaction-policy gap and the benchmark coverage that should accompany it.

## Part A: delete-bitmap density rewrite policy

The positional delete-bitmap shipped as a V1 with documented limitations: the merge-on-read rewrite is triggered only for a single-input compaction, rows are marked per-row, and superseded range-tombstones are kept verbatim. There is no policy that decides *when* a columnar segment is worth physically rewriting based on how dead it is.

Add a density-driven rewrite policy:

- Track, per columnar segment, the fraction of rows masked by its delete-bitmap (deleted-row ratio).
- When the ratio crosses a configurable threshold (e.g. 30 percent), schedule a materializing rewrite that drops the masked rows and clears the bitmap, reclaiming space and read cost. Below the threshold the bitmap is left in place (cheap MoR).
- Extend the trigger beyond the single-input case so a multi-input compaction also materializes deletions instead of copy-on-writing the bitmap forward.
- Config: `columnar_delete_density_rewrite_threshold` in `RuntimeConfig`, with an OFF mode.

## Part B: public columnar benchmark matrix

`benches/` currently has no columnar benchmark (it has `partial_decode`, `scan_since`, etc., but nothing for the PAX path). The columnar wins and overhead boundaries are therefore unproven in CI / on the bench dashboard.

Add a criterion bench matrix that quantifies:

- **Scan win**: full-column aggregate scan, columnar vs row layout, across selectivities and column counts.
- **Projection win**: scanning K of N columns, showing decode cost scales with projected columns (the projection-pushdown payoff).
- **Zone-map skip win**: range/predicate scan with vs without zone-map block skipping.
- **Delete-bitmap read cost**: scan over a segment at increasing deleted-row density, showing MoR overhead and the rewrite-threshold crossover (ties Part A to a measured number).
- **Transpose overhead boundary**: the flush/compaction-time cost of transposing into PAX, so the write-side price of columnar is explicit.

Each bench reports throughput at >= 2 scale points; results feed the public dashboard so "columnar wins, and here is where it stops winning" is demonstrable.

## Acceptance

- Density policy: a test shows a segment past the threshold gets rewritten (bitmap cleared, masked rows gone) and one below it does not; multi-input compaction materializes deletions; the threshold is configurable with an OFF mode that reproduces current behaviour.
- Benchmarks: the columnar matrix compiles, runs, and produces the five measurements above with a clear columnar-vs-row comparison; the delete-density bench shows the crossover that justifies Part A's default threshold.

## Estimate

Part A policy + trigger extension + tests ~4-5d; Part B bench matrix ~3d = ~7-8d.

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Columnar delete-bitmap density rewrite and a public columnar benchmark matrix
(the headline of #551), plus the columnar read-path performance work and two
correctness fixes surfaced along the way.

## Delete-density rewrite (#551 Part A)

A columnar segment relocated with a sub-threshold delete-bitmap (merge-on-read)
was never re-evaluated, so once its masked fraction crossed the adaptive purge
threshold (e.g. after an operator lowered it at runtime) it paid the
merge-on-read mask cost on every scan forever, with nothing scheduling the
physical rewrite.

The orchestrator now layers a density-rewrite fallback on top of the structural
strategy: when a strategy returns `DoNothing`, the densest segment whose
delete-bitmap fraction is at or above its level's `Adaptive` purge threshold is
scheduled for a single-input materializing rewrite (the masked scan drops the
dead rows; the fresh segment carries no bitmap). It self-gates to a no-op unless
a level runs `Adaptive` and holds a dense-enough live segment, applies uniformly
to every strategy without preempting real work, and reads the live runtime
config so a threshold change takes effect. `CompactionStrategy::choose` stays
purely structural.

## Columnar benchmark matrix (#551 Part B)

`benches/columnar.rs`: row vs columnar vs columnar+zonemap, across full scan and
point lookup, through the public read API, printing P50/P99/P999. It proved the
read-path gaps the rest of this PR closes.

## Columnar read-path performance

Driven by the new benchmark:

- **Scan: iterate columnar blocks directly** instead of decode -> untranspose ->
  re-encode row block -> re-parse. A new entries-backed seekable block iterator
  reads the reconstructed rows directly (bound seeks are binary searches over the
  sorted entries; MVCC version selection stays in the merge layer).
- **Scan: zero-copy untranspose.** The key column and a single non-nullable bytes
  value column are taken as shared slices, so a value larger than the inline
  threshold becomes a zero-copy view instead of a per-row allocation.
- **Point read: no whole-block re-encode.** A columnar point read decoded and
  re-encoded the entire block per lookup just to binary-search it. It now decodes
  the block once and rebuilds only the matching key's MVCC versions (binary-search
  the key column, skip masked rows) into a tiny block, then runs the normal
  seqno-aware point read. The retrieval-ribbon locator still resolves the block in
  O(1).

Measured (100k entries, no compression):

| Op | Columnar vs row-major before | after |
|----|------------------------------|-------|
| `full_scan` | ~1.67x | ~1.0x (parity) |
| `point_lookup` | ~10x | ~2.3x |

## Fixes found along the way

- **`fix(tiered)`: space-amplification debt as pending compaction bytes.** The
  size-tiered strategy returned a flat `0` from `pending_compaction_bytes`, so the
  bytes-axis write backpressure (`bytes_slowdown` / `bytes_stop`) was inert for
  tiered trees even when space amplification ran away. It now reports the bytes by
  which total L0 size exceeds its space-amplification budget. (Audit verdict: the
  L0-count axis and resume were already correct for every strategy; this was the
  one bytes-axis gap.)
- **`fix(compression)`: accept zstd negative and default levels.** The level was
  restricted to `1..=22`, and serializing a directly-constructed `Zstd(level)`
  below 1 hit a `debug_assert` panic. zstd accepts negative "fast" levels and `0`
  (= default), and our backend maps them through, so the range is now `-128..=22`
  (the on-disk format stores the level in one signed byte). Verified a negative
  level round-trips end to end.

## Testing

- Full suite: 2258 passed; columnar / point-read / MVCC suites green.
- Regression tests: tiered space-amp pending bytes; columnar density rewrite
  (a 25% segment materialized after the threshold drops to 10%); zstd negative
  level round-trip.
- `clippy --all-features --all-targets` clean; no-std check 0; doctests pass;
  `cargo doc` clean; benches compile and run.

Closes #551

## Summary by CodeRabbit

* **New Features**
  * Added an optional Criterion benchmark (`columnar` feature) for full scans and point lookups.
  * Improved columnar read paths with direct in-memory iteration and key-based point reads, including correct handling of positional deletes.
  * Added density-aware compaction fallback and exposed delete-density calculation; tiered compaction now accounts for pending “debt” for backpressure.
* **Bug Fixes**
  * Expanded accepted Zstd compression levels to the full `-128..=22` signed-byte range (including negative “fast” levels).
* **Tests**
  * Added/updated unit and integration coverage for columnar read/write edge cases, density-rewrite behavior, and compression validation.

### CODE REVIEW

File: src/table/mod.rs [hunk 1]
```diff
@@ -295,6 +295,27 @@ impl Table {
...
+    #[must_use]
+    pub fn delete_density(&self) -> Option<u8> {
+        let deleted = self.delete_bitmap().len();
+        if deleted == 0 {
+            return None;
+        }
+        let total = self.metadata.item_count.max(1);
+        // `deleted <= total`, both far below `u64::MAX / 100`, so the percentage
+        // is in `0..=100`; `.min(100)` is belt-and-suspenders for the cast.
+        let percent = (deleted * 100 / total).min(100);
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/table/mod.rs]: Fixed — the percentage is now computed in u128 (`(u128::from(deleted) * 100 / u128::from(total)).min(100)`), so the multiplication cannot overflow regardless of the inputs, then clamps to 0..=100. Zero-cost widening.

File: src/table/columnar.rs [hunk 2]
```diff
@@ -1104,6 +1104,195 @@ pub fn column_batch_to_entries(batch: &ColumnBatch) -> Result<Vec<InternalValue>
...
+    while row < row_count {
+        let k = bytes_column_row(&key_col.data, row_count, row)?;
+        if comparator.compare(k, needle) != core::cmp::Ordering::Equal {
+            break;
+        }
+        let masked = deletes.is_some_and(|(bitmap, start)| {
+            start
+                .checked_add(row)
+                .is_some_and(|pos| bitmap.contains(pos))
+        });
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/table/columnar.rs]: @coderabbitai Fixed — the masked point-read lookup now fails closed: an overflowing `block_start_row + row` position returns an error instead of treating the overflow as "not deleted", matching the scan path. Added a regression test (`column_batch_match_entries_fails_closed_when_a_delete_mask_position_overflows`).

File: src/table/mod.rs [hunk 3]
```diff
@@ -857,8 +932,9 @@ impl Table {
         for block_handle in iter {
             let block_handle = block_handle?;
 
-            // A wholly-deleted columnar block carries no keys; skip it.
-            let Some(data_block) = self.load_data_block(block_handle.as_ref())? else {
+            // A columnar block carrying no row for this key (absent / wholly
+            // deleted) returns None here; skip it.
+            let Some(data_block) = self.load_point_block(block_handle.as_ref(), key)? else {
                 continue;
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/table/mod.rs:945]: @coderabbitai Fixed — both `point_read_value` and `point_read_inner` now apply the end-key cutoff in the columnar-miss (None) branch before continuing, so an absent key below a block's end key stops the scan instead of probing later blocks.

File: tests/columnar_ingest.rs [hunk 4]
```diff
@@ -747,3 +747,52 @@ fn columnar_scan_spans_segments_with_evolving_schema() -> lsm_tree::Result<()> {
...
+            unframe_value_cells(val.as_ref(), &tags)?,
+            exp_cells,
+            "the reconstructed value unframes to the original sub-cells",
+        );
+        scanned += 1;
+    }
+    assert_eq!(
+        scanned, 2,
+        "both ingested rows are scanned through the Reconstruct path",
+    );
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/columnar_ingest.rs]: @coderabbitai Fixed — the scan test now drives the iterator explicitly and asserts it is exhausted after the expected rows, so an extra or duplicate scan row fails the test.

File: benches/columnar.rs [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    tree.flush_active_memtable(0).expect("flush");
+    (dir, tree)
+}
+
+/// Run `iters` timed invocations of `op` for Criterion's `iter_custom`, also
+/// recording each op's individual latency and printing P50/P99/P999 to stderr so
+/// a tail regression the aggregate would hide still surfaces.
+fn timed_with_tail<F: FnMut()>(label: &str, iters: u64, mut op: F) -> std::time::Duration {
+    use std::time::Instant;
+    let mut samples = Vec::with_capacity(iters as usize);
```
[the remark this answers is not in the document]
Comment (polaz) [reply benches/columnar.rs]: @coderabbitai Fixed — replaced the narrowing `iters as usize` with `usize::try_from(iters).unwrap_or(0)` (falls back to no capacity hint when it would not fit).

**Comment by polaz** [comment]
@coderabbitai review

**Comment by polaz** [comment]
@coderabbitai review
